wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -245,6 +246,72 @@ public void stop(String workspaceId) throws NotFoundException, ServerException,
     }
 
     /**
+     * Adds machine into running workspace.
+     * This method do not touch workspace configuration.
+     * Just adds machine to workspace runtime and destroy it on workspace stop.
+     * Does nothing if add already existing machine.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     * @throws ServerException
+     *         when application server is stopping
+     * @throws ConflictException
+     *         when workspace is not running
+     */
+    public void addMachine(MachineImpl machine) throws NotFoundException, ServerException, ConflictException {
```

Comment:
My issue is about using classes for a contract and not interface. (whiteboard pattern)
Machine is like 'user' view interface and in your case MachineImpl the inner model. But usually I see contracts with interface. Like Machine and any name like "ModelMachine", MachineImpl implementing this interface.

so you can use setters, contract is clear, mocking is easy. you can plug any implementation "even replace internal stuff"